Structure to match JWST Pipeline

vandalt opened the pull request:
This restructures the KPI Pipeline to match the JWST pipeline structure. The pipeline now inherits from `jwst.stpipe.Pipeline` and the steps inherit from `jwst.stpipe.Step`. This simplifies the high level logic (saving outputs, skipping steps) and makes the pipeline more uniform with regular JWST pipelines. It should also facilitate sharing steps with other pipelines.

Here is a more detailed summary of the changes brought by this PR:

- Each step now has its own directory where:
  - The `*_step.py` file contains the code for the step class.
  - Other files can contain modules for plotting or for lower level operations to make the step file more readable.
  - The step classes have been renamed to mach the JWST pipeline format, with "camel case" and ending with "Step". This should change nothing for users, as the attributes to access each step in the pipeline have not changed.
- The pipeline and all steps now inherit from JWST pipeline/step objects. This means that:
  - The `run()` method is automatically defined based on the `process()` method.
  - The attributes that were set in the `__init__` method are now set via the `spec` mechanism
  - Outputs are automatically saved after the "run" method and suffixes are set automatically
  - Print statements were replaced by the pipeline logging mechansim
- File I/O is now done in the same way as in the pipeline, i.e. using `DataModel` objects. This means that:
  - For each step, a "schema" YAML file was created to specify which FITS keywords need to be added, which types of data need to be saved in the FITS files and in which extension
  - For each step, a datamodel was created, using the YAML schema to define attributes and data structure
  - The pipeline steps now take a file name or DataModel as input to their run method, and they open it using `datamodels.open()`. Files are saved using the appropriate datamodel object for each step.
- The `constants` module now contains some constant definitions that are shared by multiple steps.

This PR changes a lot of the structure, but not functionality. All processing is still done in each step's `process()` method. I also kept the user-facing interface as untouched as possible, so that the changes under the hood to not impact end users (there might be some minor changes to do, but the `test_kpi_niriss.py` example worked almost without any changes (the changes shown by Github are mostly related to formatting for this file).

@kammerje let me know what you think. I'll update based on your comments. On my side I don't have other major updates in mind for this new structure. I'll keep using this branch in my own work and will update if I find bugs/inconsistencies with the new implementation.

EDIT: This fixes #11

vandalt committed:
Create Pipeline object and move steps to separate files

vandalt committed:
Make steps actual step objects

vandalt committed:
Move plotting to separate modules for each step

vandalt committed:
Add `plt.ioff()` at top of all plotting functions

vandalt committed:
Add datamodels specific to KPI pipeline

vandalt committed:
Use datamodel for trim step

vandalt committed:
Add test for trim step

vandalt committed:
Add kpfits datamodel class

vandalt committed:
Add all datamodels to init class

vandalt committed:
Fix jwst stsci datamodel reference paths for recenter and window

Others were already using full refrence URIs properly

vandalt committed:
Use kpfits datamodel in uncertainty step

vandalt committed:
Use datamodels in extraction step

vandalt committed:
Use datamodels to save in bad pixel step

vandalt committed:
Use datamodel to save windowing step

vandalt committed:
Use datamodels in recenter step

vandalt committed:
Use autogenerated path stem to save plots in all steps

vandalt committed:
Add todo comment about unused loop

vandalt committed:
Remove TODOs in pipeline file

vandalt committed:
Fix KP-COV dimensions in KPFITS schema (3 -> 4)

vandalt committed:
For kpfits header keys, use type `number` instead of `float64`

The latter does not work, raises an error

vandalt committed:
Rename `EmpiricalUncertainties` to `EmpiricalUncertaintiesStep`

For consitency with other steps

vandalt committed:
Add utility function to remove Kpi3 suffixes

Similar to the one for JWST models in JWST pipeline, but that knows
about KPI pipeline steps

vandalt committed:
Remove unused PathEffects from windowing steps

vandalt committed:
Fix formatting/type errors in spec of Step classes

vandalt committed:
Fix typo in empirical uncertainties plotting module name

vandalt committed:
Fix relative import level for utils and pupil models

For steps, they are now at ".." level instead of "."

vandalt committed:
Fix datamodels in bad pixel code

vandalt committed:
Fix datamodels in recenter code

vandalt committed:
Fix datamodels in windowing step

vandalt committed:
Fix datamodels in extraction step

Many things still commented out that need to be fixed/cleanup, but
committing working version (without internal recentering or windowing)

vandalt committed:
Fix datamodels in empirical uncertainties step

vandalt committed:
Add (same) custom method to remove suffix for all steps

vandalt committed:
Add test file for all steps

vandalt committed:
Use new pipeline module in `__init__.py` file

vandalt committed:
Use update steps in new pipeline module

vandalt committed:
Add test for new pipeline module

vandalt committed:
Use `kpicore` schema to share keywords between steps/datamodels

vandalt committed:
Save results in kpitest

vandalt committed:
init file for pipeline directory

vandalt committed:
Add note about guessing model type

vandalt committed:
Ignore example outputs

vandalt committed:
Propagate relevant pipeline attributes to steps

Currently, `show_plots`, `output_dir` and `good_frames` are propagated
but `good_frames` is not used by steps (need to update them)

vandalt committed:
Update and format instrument test/example scripts

vandalt committed:
Remove unused `good_frames` from empirical uncertainties step

vandalt committed:
Remove unused datamodels import from uncertainties step

vandalt committed:
Remove leftover comments

vandalt committed:
Import dq flag dictionary from jwst pipeline

Instead of hardcoding values in bad pixel step

vandalt committed:
Cleanup TODO comments

vandalt committed:
Cleanup TODO comments

vandalt committed:
Add cal_step_kpi info to mark steps completed

This enables fits headers to have "COMPLETE" step info

vandalt committed:
Move `WRAD_DEFAULT` to `constants` module

vandalt committed:
Simplify running window and recenter inside of extra step

Extract step used to take window and recenter steps as inputs to run
method and use their info to repeat steps in extraction.
Steps are still repeated, but now the attributes are passed in the
pipeline, and the core "extraction" loop is not repeated for each
scenario (value of arguments is enough to control extraction behaviour).

vandalt committed:
Add level in meta schema to have proper FITS sections for KPI

vandalt committed:
Formatting

vandalt committed:
Add `dq_mod` to window and recenter outputs

To make it flow down to KPFITS in pipeline

vandalt committed:
Change default suffix for steps that output KPFITS

vandalt committed:
Use custom `remove_suffix()` in KPI pipeline

vandalt committed:
Handle "org" and "mod" data in extraction step

vandalt committed:
Update comments in extraction step

vandalt committed:
Don't use "with" context to open datamodel in pipeline

vandalt committed:
Set KPI pipeline suffix depending on output type

vandalt committed:
Update extraction suffix in test for uncertainties

vandalt committed:
Add docstring for constants and utils

vandalt committed:
Document KPI datamodels

vandalt committed:
Document steps and pipeline

vandalt committed:
Remove previous calwebb_kpi3 module

vandalt commented:
@katherinethibault if you want to test these changes on MIRI data, any feedback is welcome!

vandalt committed:
Cleanup todo comments

vandalt commented:
As @katherinethibault pointed out offline, there are at least two things that need to be done before merging:

- [ ] Make this update compatible with 2D images (all reduction code should still work as before, but the datamodels need to be updated to accept 2D as well as 3D images)
- [ ] Either document that users must setup their environment for JWST pipeline files, or find a workaround that avoids errors from the JWST pipeline about missing environment variables

I added these two items to my TODO list and will comment here when the updates are pushed.

vandalt committed:
Add YAML schema files for 2D Image data models

vandalt committed:
Create Pyton classes for 2D Image datamodels

vandalt committed:
Handle 2D outputs in pipeline steps that produce cube or image

vandalt committed:
Override `Pipeline.finalize_results()` to avoid CRDS error.

vandalt commented:
The pipeline should now work on 2D data as well. I tested on a NIRISS `rate.fits` file. @katherinethibault you can probably test on MIRI 2D data now.

For the error raised about CRDS, it was raised by logging code that prints the CRDS context in the parent `JWSTPipeline` class. Because we don't use CRDS, the simplest option was just to override this function in the `Kpi3Pipeline` class to avoid errors when users don't have CRDS context set.

katherinethibault commented on jwst_kpi/trim_frames/trim_frames_step.py:
```diff
@@ -0,0 +1,194 @@
+    good_frames : List[int]
+        List of good frames, bad frames will be skipped.
+    trim_cent : List[int]
+        Center around which to trim (X and Y coordinates, list of length 2)
```
I don't know about the other instruments, but for MIRI the coordinates are reversed. The list of ints of trim_cent therefore looks like this: [y, x]. If this is also the case for other instruments, perhaps it would be good to specify it?

katherinethibault reviewed (approved):
The pipeline works fine for MIRI, both in 2D and 3D.

The CRDS error is also fix. I don't have to put the CRDS path anymore in my setup.

vandalt committed:
Add some important dependencies explicitely

vandalt committed:
Add webbpsf dev dependency

vandalt commented:
I converted this to draft as there are important changes (e.g. fourier bad pixels) that have been merged and that should be included here.

vandalt committed:
Merge branch 'develop' into test-merge

vandalt committed:
Update kernel phase plots with latest things from Jens

vandalt committed:
Add pupil as potential filter

vandalt committed:
Use hexa when creating KPO object

vandalt committed:
Remove loop over vectorized code

vandalt committed:
Add pupil as potential filter in recenter frames step

vandalt committed:
Use hexa when creating KPO for recenter frames

vandalt committed:
Add fourier bad pixel correction option

Implementation from Jens in previous version, copied here during merge conflict fix

vandalt commented:
I added the latest changes from `develop`, mainly the Fourier bad pixel correction, to this branch. I think it is ready for review @kammerje.

This fixes #22 automatically. I think it would also facilitate development in the future, mainly in terms of duplicate code for things like file I/O and operations shared between steps.

vandalt committed:
Fix syntax error (line not commented)

vandalt committed:
Fix `pupil_data` import level in `fourier_bpclean.py`

vandalt committed:
Add comment about reinstalling if problems with datamodels

vandalt committed:
Add NOTE comment about KPDT shape when saving kpfits
